mirgecom/mpi.py

```diff
@@ -32,6 +32,66 @@
 import sys
 
 
+def _check_gpu_oversubscription():
+    """
+    Check whether multiple ranks are running on the same GPU on each node.
+
+    Only works with pocl-cuda devices currently due to the use of the
+    PCI_DOMAIN_ID_NV extension.
+    """
+    from mpi4py import MPI
+    import pyopencl as cl
+
+    size = MPI.COMM_WORLD.Get_size()
+    rank = MPI.COMM_WORLD.Get_rank()
+
+    if size <= 1:
+        return
+
+    node_comm = MPI.COMM_WORLD.Split_type(MPI.COMM_TYPE_SHARED)
+    node_rank = node_comm.Get_rank()
+
+    cl_ctx = cl.create_some_context()
+    dev = cl_ctx.get_info(cl.context_info.DEVICES)
+
+    # No support for multi-device contexts
+    if len(dev) > 1:
+        raise RuntimeError(f"Multiple devices selected for rank {rank}.")
```

Comment:
Why? Shouldn't they be unique nonetheless?

Reply:
Fair enough. I mean, handling multi-device contexts here isn't hard (just add all their IDs before gathering), but this is OK. If you do leave it, please change the error message to `NotImplementedError("multi-device contexts not yet supported")`.

Reply:
Changed the error message.